### PR TITLE
std: refactor UNIX random data generation

### PR DESCRIPTION
**Pull request by joboet**
This PR makes a number of changes to the UNIX randomness implementation:
* Use `io::Error` for centralized error handling
* Move the file-fallback logic out of the `getrandom`-specific module
* Stop redefining the syscalls on macOS and DragonFly, they have appeared in `libc`
* Add a `OnceLock` to cache the random device file descriptor

### CODE REVIEW

File: library/std/src/sys/pal/unix/rand.rs [hunk 1]
```diff
@@ -2,35 +2,88 @@ pub fn hashmap_random_keys() -> (u64, u64) {
...
+                target_os = "solaris",
+                target_os = "illumos",
+                netbsd10,
+            ))]
+            if let Some(res) = imp::syscall(bytes) {
+                return res;
+            }
+
+            const PATH: &'static str = if cfg!(target_os = "redox") {
+                "rand:"
```
Comment (bjorn3) [inline library/std/src/sys/pal/unix/rand.rs:52]: Both `/dev/urandom` and `getrandom()` should work on Redox OS nowadays. Redox OS is currently in the process of changing the way scheme paths are represented. The new way to represent `rand:` would be `/scheme/rand`. See also https://gitlab.redox-os.org/redox-os/redox/-/issues/1431.

Comment (joboet) [reply library/std/src/sys/pal/unix/rand.rs]: I've switched it to `/dev/urandom` for now, `getrandom` isn't defined anywhere yet.

Comment (bjorn3) [reply library/std/src/sys/pal/unix/rand.rs]: I just realized that `/dev/urandom` is not available during early boot before the root filesystem is mounted. If `getrandom()` is not defined, using `/scheme/rand` is probably the best option for now.

Comment (joboet) [reply library/std/src/sys/pal/unix/rand.rs]: Alright, yeah, early boot is supposed to work. I'll change it.

File: library/std/src/sys/pal/unix/rand.rs [hunk 2]
```diff
@@ -273,39 +246,30 @@ mod imp {
...
-                panic!(
-                    "kern.arandom sysctl failed! (returned {}, s.len() {}, oldlenp {})",
-                    ret,
-                    s.len(),
-                    s_len
-                );
+            if ret == -1 {
+                return Err(Error::last_os_error());
+            } else if s_len != s.len() {
+                // FIXME(joboet): this can't actually happen, can it?
```
Comment (ChrisDenton) [inline library/std/src/sys/pal/unix/rand.rs:252]: That surprised me too. But it's pre-existing and this code will presumably disappear with time in any case so I'm not too concerned about the `FIXME` ever being addressed.

File: library/std/src/sys/pal/unix/rand.rs [hunk 3]
```diff
@@ -2,35 +2,88 @@ pub fn hashmap_random_keys() -> (u64, u64) {
...
+                return res;
+            }
+
+            const PATH: &'static str = if cfg!(target_os = "redox") {
+                "/scheme/rand"
+            } else {
+                "/dev/urandom"
+            };
+
+            static FILE: OnceLock<File> = OnceLock::new();
```
Comment (ChrisDenton) [inline library/std/src/sys/pal/unix/rand.rs:57]: Keeping the fd open and using the same one across threads seems fine on the face of it. Is this just for performance or is there some other motivation too?

**Review by ChrisDenton — APPROVED** [review]
LGTM. I've left some comments but these are just some notes to myself while I was reviewing. I do have one query but nothing that should block merging.

File: library/std/src/sys/pal/unix/rand.rs (hunk 3 above)
Comment (joboet) [reply library/std/src/sys/pal/unix/rand.rs:57]: It's mostly for performance, but it also makes the code resilient against the `/dev/urandom` device becoming unavailable due to `chroot` or things like that.

**Pull request merged**